Require applications to be referenced by name instead of ID

pb-seiso opened the pull request:
# Contributor Comments

- Added get_app_id() utility function to allow users to pass application name instead of ID
- Updated configurations to accept app_name instead of app_id
- Added unit tests for get_app_id and updated existing tests to use it

## Pull Request Checklist
Thank you for submitting a contribution to `easy_sast`!

In order to streamline the review of your contribution we ask that you review and comply with the below requirements:
 - [x] Ensure that `make build` completes successfully
 - [x] Keep our reports up-to-date by committing the output of `make report`
 - [x] Rebase your branch against the latest commit of the target branch, which likely should be `master`
 - [x] If there is an issue associated with your Pull Request, align your PR title with [this documentation](https://help.github.com/en/articles/closing-issues-using-keywords)

JonZeolla commented:
@pb-seiso merge conflict.  Let's hold on this until #36 is in, then rebase on master and take it from there.

pb-seiso commented:
@SeisoLLC/cto Rebased on master, fixed conflicts, updated reports, and ran a quick end-to-end to make sure everything still looks good. Let me know if you have any other feedback.

pb-seiso commented:
@JonZeolla Are you waiting for anything else from me before reviewing this?

JonZeolla commented:
No, sorry it's on my backlog

JonZeolla commented:
@pb-seiso please rebase on `master`

pb-seiso commented:
> @pb-seiso please rebase on `master`

Done, ready for your review.

JonZeolla commented on tests/test_api.py:
```diff
@@ -62,15 +63,16 @@ def test_veracode_xml_api_version(self):
     def test_veracode_xml_api_app_id(self):
         """
-        Test the VeracodeXMLAPI app_id property
+        Test the VeracodeXMLAPI app_name property
```
```suggestion
        Test the VeracodeXMLAPI app_id property
```

JonZeolla commented on tests/test_api.py:
```diff
@@ -62,15 +63,16 @@ def test_veracode_xml_api_version(self):
+        with patch("veracode.api.get_app_id", return_value=0):
+            veracode_xml_api = VeracodeXMLAPI("TestApp")
 
         # Fail when attempting to get the app_id property because app_id is
```
This is no longer true as of this PR.

JonZeolla commented on tests/test_api.py:
```diff
@@ -62,15 +63,16 @@ def test_veracode_xml_api_version(self):
+        Test the VeracodeXMLAPI app_name property
         """
-        veracode_xml_api = VeracodeXMLAPI()
+        with patch("veracode.api.get_app_id", return_value=0):
```
`app_id` should be a string, not an int.

pb-seiso replied:
Looks like I originally made `get_app_id` return `Union[int, None]` - I'm assuming that needs updated to `Union[str, None]` then? 

The XML schema says the app_id is a `long` which caused some confusion I think: https://analysiscenter.veracode.com/resource/2.0/applist.xsd

JonZeolla replied:
Right, https://github.com/SeisoLLC/easy_sast/blob/e220a85c7cdd2594e2f43c11f595a6d3d739175e/tests/constants.py#L49 shows my explicit test to ensure this isn't an int.  This is an example of the variation that I noted at https://github.com/SeisoLLC/easy_sast/blob/e220a85c7cdd2594e2f43c11f595a6d3d739175e/tests/constants.py#L81

JonZeolla commented on tests/constants.py:
```diff
@@ -769,6 +777,43 @@
+VALID_APPLIST_API["username"] = "TestUser"
+
+# https://help.veracode.com/reader/LMv_dtSHyb7iIxAQznC~9w/kb2SM9net26_L91VploQGw
+VALID_APPLIST_API_RESPONSE_XML = {}
```
Replicate throughout.
```suggestion
VALID_UPLOAD_API_APPLIST_RESPONSE_XML = {}
```

JonZeolla commented on tests/test_api.py:
```diff
@@ -62,15 +63,16 @@ def test_veracode_xml_api_version(self):
         """
-        veracode_xml_api = VeracodeXMLAPI()
+        with patch("veracode.api.get_app_id", return_value=0):
+            veracode_xml_api = VeracodeXMLAPI("TestApp")
```
Where possible, be specific
```suggestion
            veracode_xml_api = VeracodeXMLAPI(app_name="TestApp")
```

JonZeolla commented on tests/test_api.py:
```diff
@@ -62,15 +63,16 @@ def test_veracode_xml_api_version(self):
         self.assertRaises(ValueError, getattr, veracode_xml_api, "app_id")
 
-        # Fail when attempting to set the app_id property to an invalid value
+        # Fail when attempting to set the app_name property to an invalid value
```
```suggestion
        # Fail when attempting to set the app_id property to an invalid value
```

JonZeolla commented on tests/test_api.py:
```diff
@@ -368,7 +421,7 @@ def test_upload_api_sandbox_id(self):
         self.assertIsNone(setattr(upload_api, "sandbox_id", "12489"))
 
-        # Succeed when getting a valid app_id property
+        # Succeed when getting a valid sandbox_id property
```
Nice catch

JonZeolla commented on tests/test_config.py:
```diff
@@ -732,33 +732,48 @@ def test_apply_config(self):
-        results_api = ResultsAPI(app_id="1337")
-        applied_results_api = config.apply_config(api=results_api, config=configuration)
-        self.assertNotEqual(applied_results_api.version, applied_upload_api.version)
+        with patch("veracode.api.get_app_id", return_value="1337"):
```
See the comment below `# Note: This must be different than the above app_id`
```suggestion
        with patch("veracode.api.get_app_id", return_value="31337"):
```

JonZeolla commented on tests/test_api.py:
```diff
@@ -35,7 +35,8 @@ def test_veracode_xml_api_version(self):
         Test the VeracodeXMLAPI version property
         """
-        veracode_xml_api = VeracodeXMLAPI()
+        with patch("veracode.api.get_app_id", return_value="1337"):
```
Please do this in the 90 bazillion places across the various tests.
```suggestion
        with patch("veracode.api.get_app_id", return_value=constants.VALID_UPLOAD_API["app_id"]):
```

JonZeolla commented on veracode/api.py:
```diff
@@ -24,14 +28,16 @@ class VeracodeXMLAPI:
-    def __init__(self):
-        # Hard code these to None as they should be specified in the derived classes
-        self._app_id = None
+    def __init__(self, app_name: str):
```
Require kwarg
```suggestion
    def __init__(self, *, app_name: str):
```

JonZeolla replied:
This was set as resolved but I don't see it resolved

JonZeolla commented on veracode/api.py:
```diff
@@ -24,14 +28,16 @@ class VeracodeXMLAPI:
     """
 
-    def __init__(self):
-        # Hard code these to None as they should be specified in the derived classes
```
This is still true of `self._version = None`, reword and keep the comment

JonZeolla commented on veracode/api.py:
```diff
@@ -157,12 +188,11 @@ class UploadAPI(VeracodeXMLAPI):  # pylint: disable=too-many-instance-attributes
+    def __init__(self, *, app_name: str):
         # Don't forget to call the init of the parent
-        super().__init__()
+        super().__init__(app_name)
```
```suggestion
        super().__init__(app_name=app_name)
```

JonZeolla commented on veracode/api.py:
```diff
@@ -336,12 +366,11 @@ class ResultsAPI(VeracodeXMLAPI):
+    def __init__(self, *, app_name: str):
         # Don't forget to call the init of the parent
-        super().__init__()
+        super().__init__(app_name)
```
```suggestion
        super().__init__(app_name=app_name)
```

JonZeolla commented on veracode/api.py:
```diff
@@ -389,12 +418,10 @@ class SandboxAPI(VeracodeXMLAPI):
+    def __init__(self, *, app_name: str, sandbox_name: str):
         # Don't forget to call the init of the parent
-        super().__init__()
+        super().__init__(app_name)
```
```suggestion
        super().__init__(app_name=app_name)
```

JonZeolla commented on veracode/api.py:
```diff
@@ -24,14 +28,16 @@ class VeracodeXMLAPI:
 
+        # Set app name and look up ID
+        self._app_name = app_name
+        self._app_id = get_app_id(app_name)
```
This means we now globally require on the upload API `getapplist.do` endpoint.  Please document here, in sandbox api, results api, and in markdown.

JonZeolla replied:
Please document this somewhere

pb-seiso replied:
Note to self, make a README / wiki edit:
`Require permissions to access UploadAPI along with the other APIs, as get_app_id requires it for app ID resolution.`

JonZeolla commented on veracode/api.py:
```diff
@@ -24,14 +28,16 @@ class VeracodeXMLAPI:
         self.base_url = "https://analysiscenter.veracode.com/api/"
 
+        # Set app name and look up ID
+        self._app_name = app_name
```
Shouldn't we be using the setter to ensure it's valid?
```suggestion
        self.app_name = app_name
```

JonZeolla commented on veracode/api.py:
```diff
@@ -24,14 +28,16 @@ class VeracodeXMLAPI:
 
+        # Set app name and look up ID
+        self._app_name = app_name
+        self._app_id = get_app_id(app_name)
```
Using the setter to ensure it's valid.
```suggestion
        self.app_id = get_app_id(app_name)
```

JonZeolla commented on veracode/utils.py:
```diff
@@ -393,3 +401,41 @@ def is_valid_netloc(*, netloc: str) -> bool:
+    https://help.veracode.com/reader/orRWez4I0tnZNaA_i0zn9g/Z4Ecf1fw7868vYPVgkglww
+    """
+    try:
+        ## Use the setter to apply a default to ensure it is valid
```
I feel like this is a misplaced comment.

JonZeolla commented on veracode/utils.py:
```diff
@@ -393,3 +401,41 @@ def is_valid_netloc(*, netloc: str) -> bool:
     return False
+
+
+def get_app_id(app_name: str) -> Union[int, None]:
```
```suggestion
@validate
def get_app_id(app_name: str) -> Union[int, None]:
```

JonZeolla commented on veracode/utils.py:
```diff
@@ -393,3 +401,41 @@ def is_valid_netloc(*, netloc: str) -> bool:
+    """
+    try:
+        ## Use the setter to apply a default to ensure it is valid
+        version = {"getapplist.do": "5.0"}
```
I don't want two upload api version dicts; can we abstract this into `constants.py` for both this and `UploadAPI`'s `self.version`?  Perhaps for consistency we also move the `SandboxAPI` and `ResultsAPI` `self.version` `dict`s.

JonZeolla commented on veracode/utils.py:
```diff
@@ -393,3 +401,41 @@ def is_valid_netloc(*, netloc: str) -> bool:
+    try:
+        ## Use the setter to apply a default to ensure it is valid
+        version = {"getapplist.do": "5.0"}
+        base_url = "https://analysiscenter.veracode.com/api/"
```
Please move to `constants.py` and use here and in `VeracodeXMLAPI`'s `__init__`

JonZeolla reviewed (changes requested):
Nice work; some polish needed then should be good to go.  If you don't mind, please add my suggestions to a batch and commit them all together (if you agree with them) so I don't have two dozen new commits to sift through.

pb-seiso commented:
@JonZeolla This is ready again for review. The rebase on `master` prevented a nice commit batch and I'll have to rebase again anyway to fix commit messages and reorder them each onto `HEAD`. You can just review [this one](https://github.com/SeisoLLC/easy_sast/pull/37/commits/8af94510b7e8c3e164ea0ebc91aecfb921afebbc).

Note there will be a lint error on #43 since the commit lacks a body message. Please either fix that upstream or let me know if you want me to edit it when I rebase.

JonZeolla commented on tests/test_utils.py:
```diff
@@ -49,7 +49,13 @@ def test_results_function(
-        results_api = ResultsAPI(app_id=test_constants.VALID_RESULTS_API["app_id"])
+        with patch(
+            "veracode.api.get_app_id",
+            return_value=test_constants.VALID_UPLOAD_API["app_id"],
```
It looks like we're mixing `VALID_UPLOAD_API` and `VALID_RESULTS_API` here and below, please make sure they're aligned
```suggestion
            return_value=test_constants.VALID_RESULTS_API["app_id"],
```

pb-seiso replied:
Should be fixed now.

JonZeolla reviewed (changes requested):
Just one minor item

JonZeolla commented:
@pb-seiso this is good to go now; force pushed a commit rewording.  A simple conflict resolution is to just accept them as-is then re-generate them all.

pb-seiso commented:
@JonZeolla This is rebased and ready again for review. I ran basic integration tests on the `app_name` parameter which looked good, but please feel free to re-test.

Note that I've also included a fix to change the Docker image to one that uses Python 3.9.

JonZeolla reviewed (approved):
Nice work!  Man, is it nice to see this thing get wrapped up.  I will give you the pleasure of the squash merge; make sure to leave a properly formatted commit message for git linting.
